### PR TITLE
Pass along seed to DistributedSampler

### LINKED ISSUE #11389: Distributed DataSampler has fixed data order despite random seeds.
When using a distributed data loader with `shuffle = True` in the Hugging Face trainer, it calls the underlying torch data loader. If `shuffle` is set to True, the data loader seeds the generator with `seed + epoch` ([here](https://github.com/pytorch/pytorch/blob/f84a50109f794d4feab922056b77d7c358076776/torch/utils/data/distributed.py#L100)).

When calling the data loader in HF trainer ([here](https://github.com/huggingface/transformers/blob/3ed5e97ba04ce9b24b4a7161ea74572598a4c480/src/transformers/trainer.py#L553)), the seed is _not_ passed to the torch data loader and thereby gets set to the default seed of 0. This means the data loader generator will always gets initialized to the epoch, despite a different seed to HF.

I would think we'd want the data order to be random, too.

## Environment info

- `transformers` version: 4.5.0.dev0
- Platform: macOS-10.16-x86_64-i386-64bit
- Python version: 3.8.6
- PyTorch version (GPU?): 1.7.1 (False)
- Tensorflow version (GPU?): not installed (NA)
- Using GPU in script?: yes
- Using distributed or parallel set-up in script?: yes (with DeepSpeed)

### Who can help
@sgugger (trainer)

## Information

Model I am using (Bert, XLNet ...): GPT2

The problem arises when using:
* The hugging face trainer with a distributed data sampler

The tasks I am working on is:
* Training GPT2 from scratch using DDP with DeepSpeed

## To reproduce

Using a different seed with distributed data loader does not change the data order.

## Expected behavior

The random seed should be passed to the data loader so the data order to randomized with the seed changing.

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR passes along the seed to `DistributedSampler` otherwise it always uses 0 for setting its RNG. See #11389 for more context.

Fixes #11389

### CODE REVIEW

**Comment by siddk** [comment]
Should also be passed to the initializer for `DistributedLengthGroupedSampler`?

**Comment by sgugger** [comment]
Good point, added it!
